### ISSUE TITLE
Download a file.csv after clicking a link that opens a new tab

### ISSUE THREAD

**Issue by orliesaurus** [segmentio/nightmare#1362]
I am trying to set up a script that downloads a .csv file from a really old front-end system written in ASP.
I've spent the last 6 hours reading through stackoverflow and old issues and I finally decided to ask for help.

Using nightmarejs, I manage to login fine and navigate to the page I need, however when I try to get nightmare to "click" the anchor tag that would initiate the download of the .csv file, I am redirected to the login page of the portal.

I have inspected the anchor tag. This anchor tag has HTML content (below), and when clicked in a normal browser, opens a new tab and instantenously downloads the csv file. You can see that the onclick event initiates the download.

```html
<a href="#nowhere" onclick="App.Forms.submitForm($('reportform_csv')); return false;">Export to CSV</a>
```
I believe it's performing a POST request because when I tried to inspect with the Postman client, I have captured the data.

What can I do to overcome this issue of nightmare opening a new tab and redirecting me back to the start? 

Is there a way to directly run that onclick string in nightmarejs mimicking a user entering the string in the console in chrome `App.Forms.submitForm($('reportform_csv')); ` ? 
Because that would initiate the download and not send me back to the login page, I believe/hope?

**Comment by orliesaurus** [comment]
I have not found a solution to the .csv download problem however, by overwriting the form data using `evaluate` I was able to download the file in the same frame

**Comment by matthewmueller** [comment]
Hi @orliesaurus, that sounds like a tough problem, though it seems like you got really close. Without a runnable example, it's hard to tell exactly what the problem you encountered, but did you try [nightmare-download-manager](https://github.com/rosshinkley/nightmare-download-manager)

I've also created an example trying to mirror what you mentioned:

```js
const Nightmare = require('nightmare')
const express = require('express')
const Server = require('./server')

// setup download manager
require('nightmare-download-manager')(Nightmare)

main().catch(e => console.error('error', e))

function html(body) {
  return `<html><body>${body}</body></html>`
}

async function main() {
  const app = express()
  const server = await Server(app)

  // stream a csv
  app.post('/submit', (req, res) => {
    var data = [
      ['ID', 'Name', 'Age', 'Gender'],
      [1, 'Taro Yamada', 25, 'Male'],
      [2, 'Hanako Yamada', 24, 'Female'],
      [3, 'John Doe', 30, 'Male'],
      [4, 'Jane Doe', 30, 'Female']
    ]

    res.statusCode = 200
    res.setHeader('Content-Type', 'text/csv')
    data.forEach(function(item) {
      res.write(
        item
          .map(function(field) {
            return '"' + field.toString().replace(/\"/g, '""') + '"'
          })
          .toString() + '\r\n'
      )
    })

    res.end()
  })

  // get the form
  app.get('/', (req, res) => {
    res.send(
      html(`
      <form method="post" action="/submit">
        <input id="btn" type="submit" value="submit"/>
      </form>
    `)
    )
  })

  const nightmare = Nightmare({ show: true })
  await nightmare.downloadManager()

  nightmare.on('download', (state, downloadItem) => {
    if (state == 'started') {
      nightmare.emit('download', 'test.csv', downloadItem)
    }
  })

  const res = await nightmare.goto(server.url + '/')
  await nightmare.click('#btn')

  // wait until all downloads have completed
  await nightmare.waitDownloadsComplete()

  // shutdown
  await nightmare.end()
  await server.close()
}
```

Does this script work for your use case?

**Comment by orliesaurus** [comment]
What I did and seemed to work is this - for posterity: I was able to overwrite the HTML form found in the page in question. Changing some attributes (including adding `target="_self"`) I was able to get Nightmare download manager to work.
Basically manipulating the page's HTML I was able to get the CSV to download in the same page/view where all the browser automation happens successfully.
I have not tested your script >yet< but I believe the issue was with the fact that CSV download opened up another "tab" and lost for some reason the session & whatnot (hence why it was requesting me to login again and was stuck in a loop ever since)
